README.md
normalmap
=========

This is a command-line utility to convert PNGs to normal maps for 3D texturing.
The input is converted to a greyscale, then a 3x3 Sobel filter is applied,
which interprets the luminance of each pixel as a height, and generates a
normal map.

Installation
------------

###Linux, Unix, OS X

Make sure you have installed an appropriate toolchain for your platform,
**pkg-config** and development headers etc for **libpng** (version 1.6.0 or
later) and the **popt** library.

If the **configure** script isn't present, also install **autoconf** and
**automake**, then run `autoreconf -i`.

The compilation and installation sequence is the usual:

```
./configure
make
sudo make install
```

The **INSTALL** file contains more details about the GNU buildsystem.

###Windows

Use **normalmap.exe** from the win32 folder.

Runtime options
---------------

`normalmap --help` shows a summary of the options. Arguments to short (single
dash, single letter) options should be separated by a space, arguments for long
options should follow an equals sign wiht no spaces. Here are some details of
each option.

###--input/-i and --output/-o

Self-explanatory, these are compulsory. Input and output are both in PNG format.

###--xyz/-x

The mapping of each normal's X, Y and Z components to colour channels in the
output. The default is `rgb` ie X maps to the red channel, Y to green, and Z to
blue. You may use `a` for the alpha channel.

This program follows the convention that X and Y are axes in the plane of the
texture and Z is perpendicular to the plane. Some game engines swap Y and Z.

###--normalise/-n

The greyscale image the program uses internally has one byte per pixel, with 0
being the minimum and 255 being the maximum. If this option is given, the image
is first scanned to find the lowest/dimmest and highest/brightest pixels. All
values are then scaled to expand the range of heights to the full 255. For
example, if the lowest is 64 and the highest 191, the heights will all be
doubled so that the difference between the highest and lowest becomes 255.

No apologies for the British spelling :-).

###--scale/-s

Luminance is interpreted so that 255 corresponds to a height of 1, equal to the
width of one pixel. You can use this option to scale that height.

###--unsigned/-u

When converting the vector values back to bytes for the output pixels, the
convention is that 0 in the output corresponds to -1.0 in the vector, 128
corresponds to 0.0, and 255 corresponds to 1.0. The Z component should always
be positive, so it's possible for game engines to use a different range for Z,
where a byte of 0 means Z = 0.0 and 255 corresponds to 1.0. This option enables
that alternative Z mapping. Only use this if you're sure your game engine
supports it.

###--wrap/-w

Use for textures that are usually tiled, so that one pixel beyond the right
edge is asumed to have the same value as the pixel on the left edge at the same
Y value.

###--tm/-t

An alias for `-xyz=agb`. This makes normal maps suitable for Trackmania.

Licence
-------

Copyright (C) 2016 Tony Houghton <h@realh.co.uk>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


ChangeLog
See git <https://github.com/realh/normalmap> for the changelog.


src/main.c
/*
    normalmap - Convert PNGs from height maps to normal maps for games.
    Copyright (C) 2016 Tony Houghton <h@realh.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef _WIN32
#include <io.h>
#include <fcntl.h>
#endif

#include "nmpng.h"
#include "normalmap.h"
#include "options.h"

#ifdef _WIN32
void set_stdin_to_binary(void)
{
    _setmode(0, O_BINARY);
}

void set_stdout_to_binary(void)
{
    _setmode(1, O_BINARY);
}
#else
void set_stdin_to_binary(void)
{
    freopen(NULL, "rb", stdin);
}

void set_stdout_to_binary(void)
{
    freopen(NULL, "wb", stdout);
}
#endif

int main(int argc, char **argv)
{
    NormalmapOptions *options = normalmap_options_get(argc, argv);
    FILE *fp_in, *fp_out;
    NormalmapPng *hpng, *npng;

    /*
    fprintf(stderr, "xyz = '%s', unsigned = %d, normalise = %d\n",
            options->xyz, options->unsigned_z, options->normalise);
    */

    if (!options->input || !strcmp(options->input, "-"))
    {
        set_stdin_to_binary();
        fp_in = stdin;
    }
    else
    {
        fp_in = fopen(options->input, "rb");
        if (!fp_in)
        {
            fprintf(stderr, "Could not open input file '%s': %s\n",
                    options->input, strerror(errno));
            exit(1);
        }
    }

    if (!options->output || !strcmp(options->output, "-"))
    {
        set_stdout_to_binary();
        fp_out = stdout;
    }
    else
    {
        fp_out = fopen(options->output, "wb");
        if (!fp_out)
        {
            fprintf(stderr, "Could not open output file '%s': %s\n",
                    options->output, strerror(errno));
            exit(1);
        }
    }

    hpng = normalmap_load_png(fp_in);
    if (!hpng)
        exit(1);

    npng = normalmap_convert(hpng, options);

    normalmap_save_png(npng, fp_out);

    return 0;
}


src/nmpng.c
/*
    normalmap - Convert PNGs from height maps to normal maps for games.
    Copyright (C) 2016 Tony Houghton <h@realh.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdlib.h>
#include <string.h>

#include "nmpng.h"

NormalmapPng *normalmap_png_new(void)
{
    NormalmapPng *npng = malloc(sizeof(NormalmapPng));

    memset(npng, 0, sizeof(NormalmapPng));
    npng->info.version = PNG_IMAGE_VERSION;
    return npng;
}

NormalmapPng *normalmap_load_png(FILE *fp)
{
    NormalmapPng *npng = normalmap_png_new();

    if (!png_image_begin_read_from_stdio(&npng->info, fp))
    {
        fprintf(stderr, "Error loading PNG (begin): %s\n", npng->info.message);
        return normalmap_free_png(npng);
    }
    npng->info.format = PNG_FORMAT_GRAY;
    npng->data = malloc(npng->info.width * npng->info.height);
    if (!png_image_finish_read(&npng->info, NULL, npng->data,
            npng->info.width, NULL))
    {
        fprintf(stderr, "Error loading PNG (finish): %s\n", npng->info.message);
        return normalmap_free_png(npng);
    }
    return npng;
}

void normalmap_save_png(NormalmapPng *npng, FILE *fp)
{
    /* Make a new NormalmapPng in case resetting opaque clears other fields. */
    NormalmapPng *npng2 = normalmap_png_new();
    memcpy(npng2, npng, sizeof(NormalmapPng));
    npng2->info.opaque = NULL;
    npng2->info.flags = 0;
    if (!png_image_write_to_stdio(&npng2->info, fp, 0, npng2->data, 0, NULL))
    {
        fprintf(stderr, "Error saving PNG: %s\n", npng2->info.message);
    }
    npng2->data = NULL;
    normalmap_free_png(npng2);
}

NormalmapPng *normalmap_free_png(NormalmapPng *npng)
{
    png_image_free(&npng->info);
    free(npng->data);
    free(npng);
    return NULL;
}


src/nmpng.h
#pragma once

/*
    normalmap - Convert PNGs from height maps to normal maps for games.
    Copyright (C) 2016 Tony Houghton <h@realh.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdio.h>
#include <png.h>

typedef struct
{
    png_image info;
    unsigned char *data;
} NormalmapPng;


NormalmapPng *normalmap_load_png(FILE *fp);

void normalmap_save_png(NormalmapPng *npng, FILE *fp);

/* Returns NULL */
NormalmapPng *normalmap_free_png(NormalmapPng *npng);


src/options.h
#pragma once

/*
    normalmap - Convert PNGs from height maps to normal maps for games.
    Copyright (C) 2016 Tony Houghton <h@realh.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

typedef struct
{
    char *input;
    char *output;
    char *xyz;
    double scale;
    int unsigned_z;
    int normalise;
    int wrap;
} NormalmapOptions;

NormalmapOptions *normalmap_options_get(int argc, char **argv);


src/normalmap.c
/*
    normalmap - Convert PNGs from height maps to normal maps for games.
    Copyright (C) 2016 Tony Houghton <h@realh.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <math.h>
#include <stdlib.h>
#include <string.h>

#include "normalmap.h"

static int byte_offset(char axis)
{
    switch (axis)
    {
        case 'r':
            return 0;
        case 'g':
            return 1;
        case 'b':
            return 2;
        case 'a':
            return 3;
    }
    return 0;
}

typedef struct
{
    double x, y, z;
} NormalVector;

inline static unsigned char sample_pixel(NormalmapPng *heightmap,
        unsigned x, unsigned y)
{
    return heightmap->data[y * heightmap->info.width + x];
}

inline static int pixel_diff(NormalmapPng *heightmap,
        unsigned x1, unsigned x2, unsigned y1, unsigned y2)
{
    return (int) sample_pixel(heightmap, x2, y2) -
        (int) sample_pixel(heightmap, x1, y1);
}

static int row_dh(NormalmapPng *heightmap, unsigned x, unsigned y, int wrap)
{
    unsigned w = heightmap->info.width;

    if (x == 1)
        return 0;

    if (x == 0)
    {
        if (wrap)
        {
            return -pixel_diff(heightmap, 1, w - 1, y, y);
        }
        else
        {
            return -pixel_diff(heightmap, 0, 1, y, y);
        }
    }
    else if (x == w - 1)
    {
        if (wrap)
        {
            return -pixel_diff(heightmap, w - 2, 0, y, y);
        }
        else
        {
            return -pixel_diff(heightmap, w - 2, w - 1, y, y);
        }
    }
    /* else */
    return -pixel_diff(heightmap, x - 1, x + 1, y, y);
}

static int col_dh(NormalmapPng *heightmap, unsigned x, unsigned y, int wrap)
{
    unsigned h = heightmap->info.height;

    if (h == 1)
        return 0;

    if (y == 0)
    {
        if (wrap)
        {
            return pixel_diff(heightmap, x, x, 1, h - 1);
        }
        else
        {
            return pixel_diff(heightmap, x, x, 0, 1);
        }
    }
    else if (y == h - 1)
    {
        if (wrap)
        {
            return pixel_diff(heightmap, x, x, h - 2, 0);
        }
        else
        {
            return pixel_diff(heightmap, x, x, h - 2, h - 1);
        }
    }
    /* else */
    return pixel_diff(heightmap, x, x, y - 1, y + 1);
}

static NormalVector sobel(NormalmapPng *heightmap, unsigned x, unsigned y,
        double scale, int wrap)
{
    NormalVector nv;
    int dh;
    unsigned w = heightmap->info.width;
    unsigned h = heightmap->info.height;
    double div = 5.0;

    dh = row_dh(heightmap, x, y, wrap) * 2;
    if (y == 0)
    {
        if (wrap)
            dh += row_dh(heightmap, x, h - 1, wrap);
        else
            div -= 1.0;
    }
    else
    {
        dh += row_dh(heightmap, x, y - 1, wrap);
    }
    if (y == h - 1)
    {
        if (wrap)
            dh += row_dh(heightmap, x, 0, wrap);
        else
            div -= 1.0;
    }
    else
    {
        dh += row_dh(heightmap, x, y + 1, wrap);
    }
    nv.x = scale * (double) dh / div;

    div = 5.0;
    dh = col_dh(heightmap, x, y, wrap) * 2;
    if (x == 0)
    {
        if (wrap)
            dh += col_dh(heightmap, w - 1, y, wrap);
        else
            div -= 1.0;
    }
    else
    {
        dh += col_dh(heightmap, x - 1, y, wrap);
    }
    if (x == w - 1)
    {
        if (wrap)
            dh += col_dh(heightmap, 0, y, wrap);
        else
            div -= 1.0;
    }
    else
    {
        dh += col_dh(heightmap, x + 1, y, wrap);
    }
    nv.y = scale * (double) dh / div;

    /* Normal's Z component is always 1 before normalisation, because we
     * treat the sample points as being at the edges of the centre pixel. */
    div = sqrt(nv.x * nv.x + nv.y * nv.y + 1.0);
    nv.x /= div;
    nv.y /= div;
    nv.z = 1.0 / div;

    return nv;
}

static unsigned char d_to_signed_byte(double v)
{
    v *= 128;
    if (v == 128)
        v = 127;
    return (unsigned char) v + 128;
}

static unsigned char d_to_unsigned_byte(double v)
{
    v *= 256;
    if (v == 256)
        v = 255;
    return (unsigned char) v;
}

NormalmapPng *normalmap_convert(NormalmapPng *heightmap,
        NormalmapOptions *options)
{
    unsigned bpp;           /* bytes per pixel */
    NormalmapPng *nmap;
    int xo, yo, zo;
    unsigned int y, x, n;
    double scale;

    nmap = normalmap_png_new();
    nmap->info.width = heightmap->info.width;
    nmap->info.height = heightmap->info.height;
    if (strchr(options->xyz, 'a'))
    {
        bpp = 4;
        nmap->info.format = PNG_FORMAT_RGBA;
    }
    else
    {
        bpp = 3;
        nmap->info.format = PNG_FORMAT_RGB;
    }
    nmap->data = malloc(nmap->info.height * nmap->info.width * bpp);

    xo = byte_offset(options->xyz[0]);
    yo = byte_offset(options->xyz[1]);
    zo = byte_offset(options->xyz[2]);

    /*
     * We don't need to keep track of the minimum (offset), because we're
     * only interested in the difference between values.
     */
    if (options->normalise)
    {
        unsigned char min = 255, max = 0;
        unsigned char pix;

        x = heightmap->info.height * heightmap->info.width;
        for (y = 0; y < x; ++y)
        {
            pix = heightmap->data[y];
            if (pix > max)
                max = pix;
            if (pix < min)
                min = pix;
        }
        scale = (double) (max - min) / 255.0 * options->scale;
    }
    else
    {
        scale = options->scale / 255.0;
    }

    n = 0;
    for (y = 0; y < nmap->info.height; ++y)
    {
        for (x = 0; x < nmap->info.width; ++x)
        {
            NormalVector v = sobel(heightmap, x, y, scale, options->wrap);

            nmap->data[n + xo] = d_to_signed_byte(v.x);
            nmap->data[n + yo] = d_to_signed_byte(v.y);
            nmap->data[n + zo] = options->unsigned_z ?
                d_to_unsigned_byte(v.z) : d_to_signed_byte(v.z);

            n += bpp;
        }
    }

    return nmap;
}


src/normalmap.h
#pragma once

/*
    normalmap - Convert PNGs from height maps to normal maps for games.
    Copyright (C) 2016 Tony Houghton <h@realh.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "nmpng.h"
#include "options.h"

/*
 * heightmap is a greyscale PNG with one byte per pixel.
 * Returns a new PNG with a normal map generated from heightmap using options.
 */
NormalmapPng *normalmap_convert(NormalmapPng *heightmap,
        NormalmapOptions *options);


src/options.c
/*
    normalmap - Convert PNGs from height maps to normal maps for games.
    Copyright (C) 2016 Tony Houghton <h@realh.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <popt.h>

#include "options.h"

#ifndef PACKAGE_NAME
#define PACKAGE_NAME "normalmap"
#endif 

static struct poptOption popt_options[] =
{
    {   "input", 'i', POPT_ARG_STRING,
        NULL, 0,
        "Input file, or '-' for stdin", "FILENAME"
    },
    {   "output", 'o', POPT_ARG_STRING,
        NULL, 0,
        "Output file, or '-' for stdout", "FILENAME"
    },
    {   "xyz", 'x', POPT_ARG_STRING,
        NULL, 0,
        "Mapping of XYZ to output colour channels; eg 'rgb', 'agb'", "RGB"
    },
    {   "normalise", 'n', POPT_ARG_NONE,
        NULL, 0,
        "Scale input heightmaps values to fill range 0.0-1.0", NULL
    },
    {   "scale", 's', POPT_ARG_DOUBLE,
        NULL, 1,
        "Scale of heightmap (implies --normalise) relative to a pixel; "
        "defaults to 1.0", NULL
    },
    {   "unsigned", 'u', POPT_ARG_NONE,
        NULL, 0,
        "Z values in output are unsigned char (0-255) "
        "instead of signed (128-255)", NULL
    },
    {   "wrap", 'w', POPT_ARG_NONE,
        NULL, 0,
        "Texture wraps around for tiling", NULL
    },
    {   "tm", 't', POPT_ARG_NONE,
        NULL, 0,
        "Use Trackmania format (--xyz=agb)",
        NULL
    },
    POPT_AUTOHELP
    POPT_TABLEEND
};

static int validate_xyz(char *xyz)
{
    int m, n;

    if (strlen(xyz) != 3)
        return 0;

    for (n = 0; n < 3; ++n)
    {
        xyz[n] = tolower(xyz[n]);
    }

    if (strspn(xyz, "rgba") != 3)
        return 0;

    for (n = 0; n < 3; ++n)
    {
        for (m = 0; m < 3; ++m)
        {
            if (n != m && xyz[n] == xyz[m])
                return 0;
        }
    }

    return 1;
}

NormalmapOptions *normalmap_options_get(int argc, char **argv)
{
    poptContext pc;
    NormalmapOptions *no = malloc(sizeof(NormalmapOptions));
    char const **alias_argv = malloc(2 * sizeof(char const **));
    struct poptAlias alias_data =
    {
        "tm",
        't',
        1,
        alias_argv
    };
    int result;

    no->input = NULL;
    no->output = NULL;
    no->xyz = NULL;
    no->scale = 1.0;
    no->unsigned_z = 0;
    no->normalise = 0;

    popt_options[0].arg = &no->input;
    popt_options[1].arg = &no->output;
    popt_options[2].arg = &no->xyz;
    popt_options[3].arg = &no->normalise;
    popt_options[4].arg = &no->scale;
    popt_options[5].arg = &no->unsigned_z;
    popt_options[6].arg = &no->wrap;

    pc = poptGetContext(PACKAGE_NAME, argc, (const char **) argv,
            popt_options, 0);

    alias_argv[0] = "--xyz=agb";
    alias_argv[1] = NULL;
    poptAddAlias(pc, alias_data, 0);

    while ((result = poptGetNextOpt(pc)) >= 0)
    {
        if (result == 1)
            no->normalise = 1;
    }
    if (result < -1)
    {
        fprintf(stderr, "%s: %s\n", poptBadOption(pc, 0), poptStrerror(result));
        exit(1);
    }

    if (!no->xyz)
        no->xyz = strdup("rgb");
    if (!validate_xyz(no->xyz))
    {
        fprintf(stderr, "Bad value for --xyx: %s\n", no->xyz);
        exit(1);
    }

    if (!no->input)
    {
        fprintf(stderr, "Input must be specified\n");
        poptPrintUsage(pc, stderr, 0);
        exit(1);
    }

    if (!no->output)
    {
        fprintf(stderr, "Output must be specified\n");
        poptPrintUsage(pc, stderr, 0);
        exit(1);
    }

    poptFreeContext(pc);
    return no;
}
